design/seed_vm.md

```diff
@@ -0,0 +1,190 @@
+# SeedVM
+
+## Introduction
+
+SeedVM is a register based virtual machine which can interpret and execute the
+SeedLang byte code. Register based virtual machine avoids several push and pop
+instructions that stack based virtual machine needs to move values around the
+stack. These instructions are particularly expensive, because they involve copy
+of the values.
+
+The design intention of SeedVM is to make a stable, secure and fast virtual
+machine for the visualizable low-code programming environment that focuses on
+educational purposes.
+
+The SeedLang program is translated from SeedAST to byte code by the compiler.
+The design of the byte code is derived from the Lua 5.1. Please refer to [A
+No-Frills Introduction to Lua 5.1 VM
+Instructions](http://underpop.free.fr/l/lua/docs/a-no-frills-introduction-to-lua-5.1-vm-instructions.pdf).
+
+## SeedLang Value Types
+
+SeedLang is a dynamically typed programming language. Types are attached to
+values rather then variables. There are 9 types in SeedLang: Nil, Boolean,
+Number, String, List, Table, Function, Module and SeedObject:
+
+- Nil: Nil value, only one value nil
+- Boolean: Boolean value, true or false
+- Number: Double precision floating point number
```

Comment:
Boxing and unboxing is used in C# for transferring numbers from the stack to the heap. We don't need such kind of operation in SeedLang.
My first feeling is that we may not need the boxing and unboxing operation in current phase, because SeedLang is a dynamic programming language. All the values are objects from user's point of view. Compiler will translate the function call (3).toString() to a build-in function sys.__toString__(3). The first 3 is the object 3 in SeedAST, and the second 3 is the SeedVM value 3 with Number type. Low level implement will not impact the hight level abstract.
So I don't want to introduce the complexity of Number and number now. But I do add a description to clarify it:
```
All the SeedLang values are objects from SeedLang's perspective. They all
support properties and functions like `name`, `type`, `doc`, `toString()`, etc.
But in SeedVm, Nil, Boolean and Number are implemented by primitive number
types, other types are implemented by objects of the host language.
```

Reply:
OK.